app/controllers/benefits/CompanyBenefitController.scala

```diff
@@ -113,20 +118,40 @@ class CompanyBenefitController @Inject()(
         }
       },
       success => {
-        success match {
-          case Some(NoIDontGetThisBenefit) =>
-            journeyCacheService.cache(DecisionChoice, NoIDontGetThisBenefit) map { _ =>
-              Redirect(controllers.benefits.routes.RemoveCompanyBenefitController.stopDate())
-            }
-          case Some(YesIGetThisBenefit) =>
-            journeyCacheService.cache(DecisionChoice, YesIGetThisBenefit) map { _ =>
-              Redirect(
-                controllers.routes.ExternalServiceRedirectController
-                  .auditInvalidateCacheAndRedirectService(TaiConstants.CompanyBenefitsIform)
-                  .url)
+        val benefitType = journeyCacheService.mandatoryJourneyValueAs(
+          EndCompanyBenefit_BenefitTypeKey,
+          x => x
+        )
+
+        benefitType.flatMap {
+          case Right(name) =>
+            success match {
+              case Some(NoIDontGetThisBenefit) =>
+                journeyCacheService
+                  .cache(getBenefitDecisionKey(Some(name), DecisionChoice), NoIDontGetThisBenefit)
+                  .map { _ =>
+                    Redirect(controllers.benefits.routes.RemoveCompanyBenefitController.stopDate())
+                  }
+              case Some(YesIGetThisBenefit) =>
+                journeyCacheService.cache(getBenefitDecisionKey(Some(name), DecisionChoice), YesIGetThisBenefit).map {
+                  _ =>
+                    Redirect(controllers.routes.ExternalServiceRedirectController
+                      .auditInvalidateCacheAndRedirectService(TaiConstants.CompanyBenefitsIform))
+                }
+              case e => {
+                logger.error(s"Data in incorrect format: $e")
```

Comment:
Can you make this error message less generic?

Reply:
this case is an invalid decision being provided.

Reply:
It's not about the case but about the actual message being logged in, it's quite generic when it could be specific.

Reply:
oh, no, I get that - I've improved this (commit pending), I was just stating that it isn't a match against the benefitType.